packages/filesystem/src/browser/location/location-renderer.tsx

```diff
@@ -104,12 +192,77 @@ export class LocationListRenderer extends ReactRenderer {
         if (locationList) {
             const value = locationList.value;
             const uri = new URI(value);
-            this.service.location = uri;
+            this.trySetNewLocation(uri);
+            e.preventDefault();
+            e.stopPropagation();
+        }
+    }
+
+    protected trySetNewLocation(newLocation: URI): void {
+        if (this.lastUniqueTextInputLocation === undefined) {
+            this.lastUniqueTextInputLocation = this.service.location;
+        }
+        // prevent consecutive repeated locations from being added to location history
+        if (this.lastUniqueTextInputLocation?.path.toString() !== newLocation.path.toString()) {
+            this.lastUniqueTextInputLocation = newLocation;
+            this.service.location = newLocation;
+        }
+    }
+
+    protected async trySuggestDirectory(e: React.ChangeEvent<HTMLInputElement>): Promise<void> {
+        if (this.doAttemptAutocomplete) {
+            const locationTextInput = this.locationTextInput;
+            const { value, selectionStart } = e.currentTarget;
+            if (locationTextInput && value.slice(-1) !== '/') {
+                const valueAsURI = new URI(value);
+                const autocompleteDirectories = await this.gatherAlphabetizedDirectories(valueAsURI);
```

Comment:
My concern here is that we can spin many of those `trySuggestDirectory` async tasks. See a Theia application running with on a high-latency network:

https://user-images.githubusercontent.com/14182238/105744435-1bc36000-5f0b-11eb-9c44-e2c13aabda2b.mp4

In this case, I did `tab` when inputting `projects`, then carried on because nothing happened and when the task finished later it messed up the input. At least it's really odd.

I love the auto-completion idea, but I think we should take care of the concurrency issue here.

Reply:
I ran the following command if you want to reproduce on localhost:

```sh
sudo tc qdisk add dev lo root netem delay 300ms 10ms 25%
```

I think in this case it was a bit gruesome and you might want to reduce the high value because it takes forever for the frontend to load the way I did it...

Reply:
Ah, good catch and definitely not ideal behavior. Thanks for pointing this out. I'll look into this 👌

Reply:
Hey @marechal-p, I've implemented a caching system to handle the concurrency issue. The cache is reset whenever the text-input mode is triggered and only will request `fileService.resolve` when a new directory path is detected. The auto suggest will be rendered immediately if a cached value is available, or rendered after a slight delay if a pending promise is resolved. I've made sure to handle the scenario where the text input was being overwritten by a previously resolved promise by ensuring the resolved value is in alignment with the text input.

Thanks for the trick to slow down my machine, here is a screen capture of it under high latency:

![high-latency](https://user-images.githubusercontent.com/62660714/105917469-f52f2300-5ff7-11eb-8476-4721a2062d4a.gif)

And low latency:

![low-latency](https://user-images.githubusercontent.com/62660714/105917593-27d91b80-5ff8-11eb-8b28-a3a62baca1a0.gif)